### PR TITLE
nameof fixes

### PR DESCRIPTION
**Pull request by dsyme**
1. Allow `nameof<'T>` for name of generic parameters

2. Allow `nameof(Unchecked.defaultof<C>.Property)` for instance members

I checked and literals are already allowed:
```
> let [<Literal>] x = nameof(id);;
val x : string = "id"

> x;;
val it : string = "id"

```

I think this resolves all issues in the nameof design but @TIHan and @cartermp will check

May require fixing up some tests

### CODE REVIEW

File: src/absil/illib.fs [hunk 1]
```diff
@@ -385,6 +385,11 @@ module List =
         mn 0 xs
     let count pred xs = List.fold (fun n x -> if pred x then n+1 else n) 0 xs
 
+    let headAndTail l =
```
Comment (cartermp) [inline src/absil/illib.fs:388]: This was intentionally removed by @forki in another PR and he cleaned up (and simplified) some places where it was being used. Is it necessary to add back here?

File: src/fsharp/TypedTreeOps.fsi [hunk 2]
```diff
@@ -2279,7 +2279,11 @@ val (|EnumExpr|_|) : TcGlobals -> Expr -> Expr option
 val (|TypeOfExpr|_|) : TcGlobals -> Expr -> TType option
 
 val (|TypeDefOfExpr|_|) : TcGlobals -> Expr -> TType option
+
+val isNameOfValRef: TcGlobals -> ValRef -> bool
```
Comment (cartermp) [inline src/fsharp/TypedTreeOps.fsi:2283]: style nit: this should probably be moved down after all of the active patterns

File: tests/fsharp/core/nameof/version46/test.fsx [hunk 3]
```diff
@@ -341,68 +341,3 @@ module Foo =
     let nameof = ()
     let x = name ()
 
-do test "local variable name lookup"                    (BasicNameOfTests.``local variable name lookup`` ())
```
Comment (cartermp) [inline tests/fsharp/core/nameof/version46/test.fsx:344]: Why were these removed?

File: tests/fsharp/core/nameof/preview/test.fsx [hunk 4]
```diff
@@ -103,6 +103,14 @@ type BasicNameOfTests() =
         let b = nameof(this.MemberMethod)
         Assert.AreEqual("MemberMethod", b)
 
+    member this.``can get name of instance member via unchecked`` () =
+        let b = nameof(Unchecked.defaultof<BasicNameOfTests>.MemberMethod)
+        Assert.AreEqual("MemberMethod", b)
+
+    member this.``can get name of method type parameter``<'TTT> () =
```
Comment (cartermp) [inline tests/fsharp/core/nameof/preview/test.fsx:110]: Should also add this, perhaps as a test in another file

```fsharp
type C<'DootDoot>() =
    member _.Name() = nameof<'DootDoot>
```

File: src/absil/illib.fs (hunk 1 above)
Comment (forki) [reply src/absil/illib.fs:388]: I moved Don's cheese ;-)

**Comment by cartermp** [comment]
@dsyme what about this?

```fsharp
let rec [<Literal>] x = nameof(x)
```

File: src/absil/illib.fs (hunk 1 above)
Comment (dsyme) [reply src/absil/illib.fs:388]: Yeah this is a useful function :-)

File: tests/fsharp/core/nameof/version46/test.fsx (hunk 3 above)
Comment (dsyme) [reply tests/fsharp/core/nameof/version46/test.fsx:344]: The test is for generating errors, not for execution, I find it confusing if they're there but not actually being executed

File: tests/fsharp/core/nameof/preview/test.fsx (hunk 4 above)
Comment (dsyme) [reply tests/fsharp/core/nameof/preview/test.fsx:110]: Yes agreed

**Comment by cartermp** [comment]
The literal business is "resolved" insofar as it requires a langauge suggestion
